flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/remote/result/TableResultWrapper.java

```diff
@@ -0,0 +1,175 @@
+    }
+
+    public String getResultId() {
+        return resultId;
+    }
+
+    public void setMaterialized(boolean isMaterialized) {
+        this.isMaterialized = isMaterialized;
+    }
+
+    public boolean isMaterialized() {
+        return isMaterialized;
+    }
+
+    public void setConfig(ReadableConfig config) {
+        this.config = config;
+    }
+
+    public ReadableConfig getConfig() {
+        return config;
+    }
+
+    /** Cannot get job client through SQL Gateway. */
+    @Override
+    public Optional<JobClient> getJobClient() {
+        return Optional.empty();
+    }
+
+    @Override
+    public void await() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void await(long timeout, TimeUnit unit) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public ResolvedSchema getResolvedSchema() {
+        return resolvedSchema;
+    }
+
+    @Override
+    public ResultKind getResultKind() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public CloseableIterator<Row> collect() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void print() {
+        throw new UnsupportedOperationException();
+    }
+
+    /** Returns an iterator that returns the iterator with the internal row data type. */
+    @Override
+    public CloseableIterator<RowData> collectInternal() {
+        return dataIterator;
+    }
+
+    @Override
+    public RowDataToStringConverter getRowDataToStringConverter() {
+        // todo
+        return rowData -> new String[] {"FAKE TEST RETURN"};
+    }
+
+    // --------------------------------------------------------------------------------------------
+
+    private static class RowDataIterator implements CloseableIterator<RowData> {
+
+        private final RemoteExecutor executor;
+        private final OperationHandle operationHandle;
+        private Iterator<RowData> currentData;
+
+        private Long nextToken;
+
+        public RowDataIterator(
+                RemoteExecutor executor,
+                OperationHandle operationHandle,
+                List<RowData> data,
+                Long nextToken) {
+            this.executor = executor;
+            this.operationHandle = operationHandle;
+            this.currentData = data.iterator();
+            this.nextToken = nextToken;
+        }
+
+        @Override
+        public void close() {
+            // do nothing
+        }
+
+        // todo: not sure whether if this is correct
+        @Override
+        public boolean hasNext() {
+            while (!currentData.hasNext() && nextToken != null) {
+                FetchResultsResponseBody response =
+                        executor.fetchResults(operationHandle, nextToken);
+                currentData = response.getResults().getData().iterator();
+                nextToken = executor.parseTokenFromUri(response.getNextResultUri());
```

Comment:
Before results are ready, the response will be an empty description of the results (e.g., there is a schema returned).  

This loop needs some kind of back-off / wait mechanism. 

As a quick example, I did something like this locally:

```                if (!currentData.hasNext() && nextToken != null) {
                    try {
                        Thread.sleep(200);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }```

Here `200` is a magic value; I'd suggest that we externalize any polling time as a configuration.